Support Sourcify in `hardhat-verify`

manuelwedler opened the pull request:
<!--
Thank you for using Hardhat and taking the time to send a Pull Request!

If you are introducing a new feature, please discuss it in an Issue or with someone from the team before submitting your change.

Please:
 - consider the checklist items below
 - keep the ones that make sense for your PR, and
 - DELETE the items that DON'T make sense for your PR.

## Note about small PRs and airdrop farming

We generally really appreciate external contributions, and strongly encourage meaningful additions and fixes! However, due to a recent increase in small PRs potentially created to farm airdrops, we might need to close a PR without explanation if any of the following apply:

- It is a change of very minor value that still requires additional review time/fixes (e.g. PRs fixing trivial spelling errors that can’t be merged in less than a couple of minutes due to incorrect suggestions)
- It introduces inconsequential changes (e.g. rewording phrases)
- The author of the PR does not respond in a timely manner
- We suspect the Github account of the author was created for airdrop farming
-->

- [ ] Because this PR includes a **bug fix**, relevant tests have been included.
- [x] Because this PR includes a **new feature**, the change was previously discussed on an Issue or with someone from the team.
- [ ] I didn't do anything of this.

---

<!-- Add a description of your PR here -->

Closes #6885

Related documentation change: https://github.com/NomicFoundation/hardhat-website/pull/116

### Changes

- Adds the `Sourcify` class which implements the `VerificationProvider`
- Sourcify is enabled by default, and called from the main verify task
- Also adds the Sourcify subtask
- Adds an additional `--creation-tx-hash` option to the verify task, which can optionally be provided by the user to more easily get a creation match
- Sourcify URLs can be configured via the user config of the plugin. Note that Etherscan urls are configured via the global chain descriptors, but I think Sourcify doesn't fit into being added as an explorer.
- I had to change how the compiler input is passed to the `verify` function of the `VerificationProvider` interface, because the Sourcify API expects the non-stringified version of the compiler input.

### Example

<img width="1014" height="256" alt="Screenshot from 2025-10-21 15-10-00_0" src="https://github.com/user-attachments/assets/fa7a00a1-c116-425e-ad6a-3f6f32c8f124" />

schaable commented on v-next/hardhat-verify/src/internal/tasks/verify/types.ts:
```diff
@@ -2,11 +2,12 @@ import type { LibraryAddresses } from "../../libraries.js";
 export interface VerifyActionArgs {
   address: string;
-  constructorArgs: string[];
+  constructorArgs?: string[];
```
Let's keep `constructorArgs` required for now, and Sourcify can just ignore them. We can update this later if users request it.

manuelwedler replied:
sounds good

schaable commented on v-next/hardhat-verify/src/internal/sourcify.ts:
```diff
@@ -0,0 +1,376 @@
+const VERIFICATION_STATUS_POLLING_SECONDS = 3;
+
+export const SOURCIFY_REPO_URL = "https://repo.sourcify.dev";
+export const SOURCIFY_API_URL = "https://sourcify.dev/server";
```
Are the staging and local urls commonly used? How often do users use a different repo or api url?

manuelwedler commented on v-next/hardhat-utils/src/internal/request.ts:
```diff
@@ -30,14 +30,14 @@ export async function generateTempFilePath(filePath: string): Promise<string> {
 export async function getBaseRequestOptions(
   requestUrl: string,
-  { extraHeaders, abortSignal, queryParams }: RequestOptions = {},
+  { extraHeaders, abortSignal, queryParams, throwOnError }: RequestOptions = {},
```
Okay, I see. I can make the logic work without it.

schaable replied:
I think you need to specify the response body in the interceptor. Once everything else is ready, push the code and I’ll take a look at the tests.

manuelwedler replied:
Thanks. The pushed commits should already solve everything else.

manuelwedler commented:
Btw, I added two additional changes:

With a recent update of the Sourcify server, it is now possible to reach the repo ui via redirect from the API. Therefore, I added a change to remove the repo URL from the configuration. This reduces the configuration overhead a bit.
If you disagree with the change, we can move back to have two configuration variables - `url` and `apiUrl`.

I also added a commit to show a link to the job info page of the Sourcify UI in case the verification fails. The page provides more info to the user for debugging why the verification fails. See for example: https://verify.sourcify.dev/jobs/75d1a316-dd30-4fb1-a889-e6427e3d3211

schaable commented:
Thanks, @manuelwedler! I’ll do another pass to fix any minor details I find along with the tests. After that, we’re good to go.

Regarding:

> With a recent update of the Sourcify server, it is now possible to reach the repo ui via redirect from the API. Therefore, I added a change to remove the repo URL from the configuration. This reduces the configuration overhead a bit.
> If you disagree with the change, we can move back to have two configuration variables - url and apiUrl.

Is this also available for custom Sourcify instances like https://sourcify-api-monad.blockvision.org/health
 or https://verify-api.vechainstats.com/health ?
